</br></br>

### TREND-DB

**TREND-DB** is a database that visualizes the **dynamics of alternative polyadenylation (APA)** influenced by various co- and post-transcriptional events.

**TREND-DB** allows to interactively explore the dynamic landscape of APA events caused by siRNA mediated downregulation of 174 APA-regulators (targeting various facets of transcriptional, co- and posttranscriptional gene regulation, epigenetic modifications and other categories) in Neuroblastoma BE(2)-C cells. Further entities are in progress.

The database facilitates:

- querying genes affected by specific APA-regulators
- querying APA-regulators affecting specific genes (with a graphical illustration of global genome-wide APA-effects)
- GO (Gene Ontology) enrichment of APA-affected genes
- visualization of APA-effects on internal gene visualization tools and on UCSC Genome Browser
- visualization of additional layers of gene regulation accounting for individual APA-affected target RNAs (*i.e.* miRNA binding sites) 

### News

- **September 2020** - The manuscript for TREND-DB ("TREND-DB—a transcriptome-wide atlas of the dynamic landscape of alternative polyadenylation") is out on Nucleic Acids Research, and will be included in the Database Issue. You can read it <a href="https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gkaa722/5911742" target="_blank">here</a>, or check out its preprint on <a href="https://www.biorxiv.org/content/10.1101/2020.08.04.235804v2" target="_blank">bioRxiv</a>!

- **August 2020** - TREND-DB is released in its version 1.0.0.

- **December 2018** - The PCF11 paper is out on <a href="https://www.nature.com/articles/s41467-018-07580-5"  target="_blank">Nature Communications</a> and has been chosen as <a href="https://www.nature.com/collections/bhkhbxstgj" target="_blank">Editor Highlight</a>!
- **October 2018** - The paper about PCF11 as critical regulator of transcriptome-3'end-diversification is out on [biorXiv](https://www.biorxiv.org)! You can read it <a href="https://www.biorxiv.org/content/early/2018/10/18/426536" target="_blank">here</a>.

### Citations

- **TRENDseq - A highly multiplexed high throughput 3’end RNA sequencing for mapping alternative polyadenylation.** 
  *Ogorodnikov A, Danckwardt S.*
  Methods in Enzymology, doi: https://doi.org/10.1016/bs.mie.2021.03.022
- **Transcriptome 3′end organization by PCF11 links alternative polyadenylation to formation and neuronal differentiation of neuroblastoma**
  *Anton Ogorodnikov, Michal Levin, Surendra Tattikota, Sergey Tokalov, Mainul Hoque, Denise Scherzinger, Federico Marini, Ansgar Poetsch, Harald Binder, Stephan Macher-Göppinger, Hans Christian Probst, Bin Tian, Michael Schaefer, Karl J. Lackner, Frank Westermann & Sven Danckwardt*.
  Nature Communications volume 9, Article number: 5331 (2018); doi: https://doi.org/10.1038/s41467-018-07580-5
- **PCF11 connects alternative polyadenylation to formation and spontaneous regression of neuroblastoma**
  *Anton Ogorodnikov, Michal Levin, Surendra Tattikota, Sergey Tokalov, Mainul Hoque, Denise Scherzinger, Federico Marini, Ansgar Poetsch, Harald Binder, Stephan Macher-Goeppinger, Bin Tian, Michael Schaefer, Karl Lackner, Frank Westermann, Sven Danckwardt*. 
  bioRxiv 426536; doi: https://doi.org/10.1101/426536
- **Processing and transcriptome expansion at the mRNA 3′ end in health and disease: finding the right end**
  *Ogorodnikov, A., Kargapolova, Y. & Danckwardt, S.*. Pflugers Arch - Eur J Physiol (2016) 468: 993.
  https://doi.org/10.1007/s00424-016-1828-3

### Responsible groups </br>

- Danckwardt Lab @ University Medical Center Mainz (<a href="https://www.unimedizin-mainz.de/cth/professuren-arbeitsgruppen/prof-danckwardt.html" target="_blank">link</a>) 
- IMBEI - Division Biostatistics and Bioinformatics @ University Medical Center Mainz (<a href="http://www.unimedizin-mainz.de/imbei/biostatistics-bioinformatics/welcome.html?L=1" target="_blank">link</a>) 


### How to use **TREND-DB**

The latest release of TREND-DB features a dynamic tour, implemented via the `rintrojs` package.
These allow the user to take a guided tour of the interface, highlighting elements that can be interacted with - by doing so, it is possible to showcase a typical use case, and at the same time the user can familiarize with the functionality of TREND-DB.

For general information, please refer to the content of the **About** tab.

### Further reading

- <a href="https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gkaa722/5911742" target="_blank">Marini et al. Nucleic Acids Research 2020</a>
- <a href="https://doi.org/10.1016/bs.mie.2021.03.022" target="_blank">Ogorodnikov et al. Methods in Enzymology 2021</a>
- <a href="https://www.mdpi.com/2218-273X/10/6/915" target="_blank">Nourse et al. Biomolecules 2020</a>
- <a href="https://www.nature.com/articles/s41467-018-07580-5" target="_blank">Ogorodnikov et al. Nature Communications 2018</a>
- <a href="http://doi.org/10.1007/s00424-016-1828-3" target="_blank">Ogorodnikov et al. Pflugers Arch. 2016</a>
- <a href="http://doi.org/10.1038/nrm.2016.116" target="_blank">Tian & Manley Nat Rev Mol Cell Biol 2016</a> 
- <a href="http://doi.org/10.1038/nrg3482" target="_blank">Elkon et al. Nat Rev Genet 2013</a>

